kotlin/generate-geodatabase/src/main/java/com/esri/arcgisruntime/sample/generategeodatabase/MainActivity.kt

```diff
@@ -0,0 +1,200 @@
+
+class MainActivity : AppCompatActivity() {
+
+  private val TAG = MainActivity::class.java.simpleName
+  // define the local path where the geodatabase will be stored
+  private val localGeodatabasePath = cacheDir.path + getString(R.string.wildfire_geodatabase)
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+    // use local tile package for the base map
+    val sanFrancisco =
+      TileCache(getExternalFilesDir(null).toString() + getString(R.string.san_francisco_tpk))
+    val tiledLayer = ArcGISTiledLayer(sanFrancisco)
+
+    // add the map and graphics overlay to the map view
+    mapView.apply {
+      // create a map with the tile package basemap
+      this.map = ArcGISMap(Basemap(tiledLayer))
+      // create a graphics overlay to display the boundaries
+      graphicsOverlays.add(GraphicsOverlay())
+    }
+  }
+
+  /**
+   * Creates a GenerateGeodatabaseJob and runs it.
+   *
+   * @param view the button which calls this function
+   */
+  fun generateGeodatabase(view: View) {
+    // create a geodatabase sync task and load it
+    val geodatabaseSyncTask = GeodatabaseSyncTask(getString(R.string.wildfire_sync))
+    geodatabaseSyncTask.loadAsync()
+
+    geodatabaseSyncTask.addDoneLoadingListener {
+      // draw a box around the extent
+      mapView.apply {
+        // clear any previous operational layers and graphics
+        map.operationalLayers.clear()
+        graphicsOverlays[0].graphics.clear()
+        // show the extent used as a graphic
+        graphicsOverlays[0].graphics.add(
+          Graphic(
+            visibleArea.extent,
+            SimpleLineSymbol(SimpleLineSymbol.Style.SOLID, Color.RED, 5f)
+          )
+        )
+      }
+
+      // create parameters for the job with the return attachments option set to false
+      val parameters = geodatabaseSyncTask
+        .createDefaultGenerateGeodatabaseParametersAsync(mapView.visibleArea.extent).get()
+        .apply { isReturnAttachments = false }
+
+      // create the generate geodatabase job
+      val generateGeodatabaseJob =
+        geodatabaseSyncTask.generateGeodatabase(parameters, localGeodatabasePath)
+
+      // show the job's progress in a dialog
+      val dialog = createProgressDialog(generateGeodatabaseJob)
+      dialog.show()
+      // start the job
+      generateGeodatabaseJob.start()
+      // update progress
+      generateGeodatabaseJob.addProgressChangedListener {
+        dialog.progressBar.progress = generateGeodatabaseJob.progress
+        dialog.progressTextView.text = "$generateGeodatabaseJob.progress%"
+      }
+      // get geodatabase when done
+      generateGeodatabaseJob.addJobDoneListener {
+        // close the progress dialog
+        dialog.dismiss()
+        // load the geodatabase and display its feature tables on the map
+        loadGeodatabase(generateGeodatabaseJob, geodatabaseSyncTask)
+      }
```

Comment:
```suggestion
      // define progress and done behaviours and start the job
      generateGeodatabaseJob.apply {
        // update progress
        addProgressChangedListener {
          dialog.progressBar.progress = this.progress
          dialog.progressTextView.text = "$this.progress%"
        }
        // get geodatabase when done
        addJobDoneListener {
          // close the progress dialog
          dialog.dismiss()
          // load the geodatabase and display its feature tables on the map
          loadGeodatabase(generateGeodatabaseJob, geodatabaseSyncTask)
        }
      }.start()
```

Reply:
I think it's the listeners should probably still be in an apply block since they're building out behaviour of the job. Just think the call to start shouldn't be in the apply block, but keen to know if you think this is still readable

Reply:
yeah, I think this is readable!